cffi backend is hanging in ipykernel tests on PyPy

The [changelog for 20.0.0](https://github.com/zeromq/pyzmq/blob/master/docs/source/changelog.rst) mention that you would like to get feedback on the new cffi backend, so here goes :). Of course I am willing to help solve this, especially if it is specific to PyPy.

Over in the conda-forge ipykernel-feedstock](https://github.com/conda-forge/ipykernel-feedstock/pull/72) (which uses pyzmq under the hood) there is a strange hang in the PyPy runs. When I ctrl-c in the hanging test I get `/home/matti/miniconda3/envs/pypy37/site-packages/zmq/backend/cffi/_poll.py:56: KeyboardInterrupt` which makes me think it is somewhere in the new cffi backend. Here are the admittedly complicated steps to reproduce

Is there an easy way to try this with cpython and enable the cffi backend?

```
conda create -n pypy37 pypy3.7
conda activate pypy37
conda install "ipython>=5.0" jupyter_client tornado traitlets curio flaky numpy pytest nose
wget https://pypi.io/packages/source/i/ipykernel/ipykernel-5.4.2.tar.gz
tar -xf ipykernel-5.4.2.tar.gz
cd ipykernel-5.4.2/
python setup.py bdist_wheel
python -m pip install dist/ipykernel-5.4.2-py3-none-any.whl 
python -m ipykernel install --sys-prefix
jupyter kernelspec list
# Note the output, which for me is 
# python3    /home/matti/miniconda3/envs/pypy37/share/jupyter/kernels/python3
# Take the prefix before the "share" for the next step
export PREFIX=/home/matti/miniconda3/envs/pypy37
$ python -m pytest -vv -m "not flaky" --full-trace ipykernel/tests/
```

and the (huge) full traceback

<details>

```
$ python -m pytest -vv -m "not flaky" --full-trace ipykernel/tests/
====================================================== test session starts =======================================================
platform linux -- Python 3.7.9[pypy-7.3.3-beta], pytest-6.2.1, py-1.10.0, pluggy-0.13.1 -- /home/matti/miniconda3/envs/pypy37/bin/python
cachedir: .pytest_cache
rootdir: /home/matti/pypy_stuff/conda-feedstocks/feedstocks/ipykernel/ipykernel-5.4.2
plugins: flaky-3.7.0
collected 103 items                                                                                                              

ipykernel/tests/test_async.py::test_async_await ^C

======================================================== warnings summary ========================================================
../../../../../miniconda3/envs/pypy37/site-packages/nose/importer.py:12
  /home/matti/miniconda3/envs/pypy37/site-packages/nose/importer.py:12: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    from imp import find_module, load_module, acquire_lock, release_lock

ipykernel/zmqshell.py:57
  /home/matti/pypy_stuff/conda-feedstocks/feedstocks/ipykernel/ipykernel-5.4.2/ipykernel/zmqshell.py:57: DeprecationWarning: ipykernel.datapub is deprecated. It has moved to ipyparallel.datapub
    from ipykernel.datapub import ZMQDataPublisher

ipykernel/datapub.py:21
  /home/matti/pypy_stuff/conda-feedstocks/feedstocks/ipykernel/ipykernel-5.4.2/ipykernel/datapub.py:21: DeprecationWarning: ipykernel.serialize is deprecated. It has moved to ipyparallel.serialize
    from ipykernel.serialize import serialize_object

ipykernel/serialize.py:25
  /home/matti/pypy_stuff/conda-feedstocks/feedstocks/ipykernel/ipykernel-5.4.2/ipykernel/serialize.py:25: DeprecationWarning: ipykernel.pickleutil is deprecated. It has moved to ipyparallel.
    from ipykernel.pickleutil import (

ipykernel/pickleutil.py:26
  /home/matti/pypy_stuff/conda-feedstocks/feedstocks/ipykernel/ipykernel-5.4.2/ipykernel/pickleutil.py:26: DeprecationWarning: ipykernel.codeutil is deprecated since IPykernel 4.3.1. It has moved to ipyparallel.serialize
    from ipykernel import codeutil

-- Docs: https://docs.pytest.org/en/stable/warnings.html
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

config = <_pytest.config.Config object at 0x00007f92e31baf38>, doit = <function _main at 0x00007f92e2dbf560>

    def wrap_session(
        config: Config, doit: Callable[[Config, "Session"], Optional[Union[int, ExitCode]]]
    ) -> Union[int, ExitCode]:
        """Skeleton command line program."""
        session = Session.from_config(config)
        session.exitstatus = ExitCode.OK
        initstate = 0
        try:
            try:
                config._do_configure()
                initstate = 1
                config.hook.pytest_sessionstart(session=session)
                initstate = 2
>               session.exitstatus = doit(config, session) or 0

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/main.py:269: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

config = <_pytest.config.Config object at 0x00007f92e31baf38>
session = <Session ipykernel-5.4.2 exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=103>

    def _main(config: Config, session: "Session") -> Optional[Union[int, ExitCode]]:
        """Default command line protocol for initialization, session,
        running tests and reporting."""
        config.hook.pytest_collection(session=session)
>       config.hook.pytest_runtestloop(session=session)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/main.py:323: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_HookCaller 'pytest_runtestloop'>, args = ()
kwargs = {'session': <Session ipykernel-5.4.2 exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=103>}, notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.spec and self.spec.argnames:
            notincall = (
                set(self.spec.argnames) - set(["__multicall__"]) - set(kwargs.keys())
            )
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call".format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self.get_hookimpls(), kwargs)

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/hooks.py:286: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.config.PytestPluginManager object at 0x00007f92e228d948>, hook = <_HookCaller 'pytest_runtestloop'>
methods = [<HookImpl plugin_name='main', plugin=<module '_pytest.main' from '/home/matti/miniconda3/envs/pypy37/site-packages/_p...in.py'>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x00007f92e1f70b48>>]
kwargs = {'session': <Session ipykernel-5.4.2 exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=103>}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/manager.py:93: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook = <_HookCaller 'pytest_runtestloop'>
methods = [<HookImpl plugin_name='main', plugin=<module '_pytest.main' from '/home/matti/miniconda3/envs/pypy37/site-packages/_p...in.py'>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x00007f92e1f70b48>>]
kwargs = {'session': <Session ipykernel-5.4.2 exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=103>}

    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
        methods,
        kwargs,
>       firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
    )

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/manager.py:87: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

session = <Session ipykernel-5.4.2 exitstatus=<ExitCode.OK: 0> testsfailed=0 testscollected=103>

    def pytest_runtestloop(session: "Session") -> bool:
        if session.testsfailed and not session.config.option.continue_on_collection_errors:
            raise session.Interrupted(
                "%d error%s during collection"
                % (session.testsfailed, "s" if session.testsfailed != 1 else "")
            )
    
        if session.config.option.collectonly:
            return True
    
        for i, item in enumerate(session.items):
            nextitem = session.items[i + 1] if i + 1 < len(session.items) else None
>           item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/main.py:348: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_HookCaller 'pytest_runtest_protocol'>, args = ()
kwargs = {'item': <Function test_async_await>, 'nextitem': <Function test_async_interrupt[asyncio]>}, notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.spec and self.spec.argnames:
            notincall = (
                set(self.spec.argnames) - set(["__multicall__"]) - set(kwargs.keys())
            )
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call".format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self.get_hookimpls(), kwargs)

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/hooks.py:286: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.config.PytestPluginManager object at 0x00007f92e228d948>, hook = <_HookCaller 'pytest_runtest_protocol'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/home/matti/miniconda3/envs/pypy37/site-package...ings', plugin=<module '_pytest.warnings' from '/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/warnings.py'>>]
kwargs = {'item': <Function test_async_await>, 'nextitem': <Function test_async_interrupt[asyncio]>}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/manager.py:93: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook = <_HookCaller 'pytest_runtest_protocol'>
methods = [<HookImpl plugin_name='runner', plugin=<module '_pytest.runner' from '/home/matti/miniconda3/envs/pypy37/site-package...ings', plugin=<module '_pytest.warnings' from '/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/warnings.py'>>]
kwargs = {'item': <Function test_async_await>, 'nextitem': <Function test_async_interrupt[asyncio]>}

    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
        methods,
        kwargs,
>       firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
    )

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/manager.py:87: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <flaky.flaky_pytest_plugin.FlakyPlugin object at 0x00007f92e1fe23a0>, item = <Function test_async_await>
nextitem = <Function test_async_interrupt[asyncio]>

    def pytest_runtest_protocol(self, item, nextitem):
        """
        Pytest hook to override how tests are run.
    
        Runs a test collected by pytest.
        - First, monkey patches the builtin runner module to call back to
        FlakyPlugin.call_runtest_hook rather than its own.
        - Then defers to the builtin runner module to run the test,
        and repeats the process if the test needs to be rerun.
        - Reports test results to the flaky report.
    
        :param item:
            pytest wrapper for the test function to be run
        :type item:
            :class:`Function`
        :param nextitem:
            pytest wrapper for the next test function to be run
        :type nextitem:
            :class:`Function`
        :return:
            True if no further hook implementations should be invoked.
        :rtype:
            `bool`
        """
        test_instance = self._get_test_instance(item)
        self._copy_flaky_attributes(item, test_instance)
        if self.force_flaky and not self._has_flaky_attributes(item):
            self._make_test_flaky(
                item,
                self.max_runs,
                self.min_passes,
            )
        original_call_and_report = self.runner.call_and_report
        self._call_infos[item] = {}
        should_rerun = True
        try:
            self.runner.call_and_report = self.call_and_report
            while should_rerun:
>               self.runner.pytest_runtest_protocol(item, nextitem)

/home/matti/miniconda3/envs/pypy37/site-packages/flaky/flaky_pytest_plugin.py:94: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <Function test_async_await>, nextitem = <Function test_async_interrupt[asyncio]>

    def pytest_runtest_protocol(item: Item, nextitem: Optional[Item]) -> bool:
        ihook = item.ihook
        ihook.pytest_runtest_logstart(nodeid=item.nodeid, location=item.location)
>       runtestprotocol(item, nextitem=nextitem)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/runner.py:109: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <Function test_async_await>, log = True, nextitem = <Function test_async_interrupt[asyncio]>

    def runtestprotocol(
        item: Item, log: bool = True, nextitem: Optional[Item] = None
    ) -> List[TestReport]:
        hasrequest = hasattr(item, "_request")
        if hasrequest and not item._request:  # type: ignore[attr-defined]
            item._initrequest()  # type: ignore[attr-defined]
>       rep = call_and_report(item, "setup", log)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/runner.py:120: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <flaky.flaky_pytest_plugin.FlakyPlugin object at 0x00007f92e1fe23a0>, item = <Function test_async_await>, when = 'setup'
log = True, kwds = {}

    def call_and_report(self, item, when, log=True, **kwds):
        """
        Monkey patched from the runner plugin. Responsible for running
        the test and reporting the outcome.
        Had to be patched to avoid reporting about test retries.
    
        :param item:
            pytest wrapper for the test function to be run
        :type item:
            :class:`Function`
        :param when:
            The stage of the test being run. Usually one of 'setup', 'call', 'teardown'.
        :type when:
            `str`
        :param log:
            Whether or not to report the test outcome. Ignored for test
            retries; flaky doesn't report test retries, only the final outcome.
        :type log:
            `bool`
        """
>       call = call_runtest_hook(item, when, **kwds)

/home/matti/miniconda3/envs/pypy37/site-packages/flaky/flaky_pytest_plugin.py:138: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <Function test_async_await>, when = 'setup', kwds = {}
reraise = (<class '_pytest.outcomes.Exit'>, <class 'KeyboardInterrupt'>)

    def call_runtest_hook(
        item: Item, when: "Literal['setup', 'call', 'teardown']", **kwds
    ) -> "CallInfo[None]":
        if when == "setup":
            ihook: Callable[..., None] = item.ihook.pytest_runtest_setup
        elif when == "call":
            ihook = item.ihook.pytest_runtest_call
        elif when == "teardown":
            ihook = item.ihook.pytest_runtest_teardown
        else:
            assert False, f"Unhandled runtest hook case: {when}"
        reraise: Tuple[Type[BaseException], ...] = (Exit,)
        if not item.config.getoption("usepdb", False):
            reraise += (KeyboardInterrupt,)
        return CallInfo.from_call(
>           lambda: ihook(item=item, **kwds), when=when, reraise=reraise
        )

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/runner.py:255: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class '_pytest.runner.CallInfo'>, func = <function call_runtest_hook.<locals>.<lambda> at 0x000056409e576340>
when = 'setup', reraise = (<class '_pytest.outcomes.Exit'>, <class 'KeyboardInterrupt'>)

    @classmethod
    def from_call(
        cls,
        func: "Callable[[], TResult]",
        when: "Literal['collect', 'setup', 'call', 'teardown']",
        reraise: Optional[
            Union[Type[BaseException], Tuple[Type[BaseException], ...]]
        ] = None,
    ) -> "CallInfo[TResult]":
        excinfo = None
        start = timing.time()
        precise_start = timing.perf_counter()
        try:
>           result: Optional[TResult] = func()

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/runner.py:311: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>       lambda: ihook(item=item, **kwds), when=when, reraise=reraise
    )

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/runner.py:255: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_HookCaller 'pytest_runtest_setup'>, args = (), kwargs = {'item': <Function test_async_await>}, notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.spec and self.spec.argnames:
            notincall = (
                set(self.spec.argnames) - set(["__multicall__"]) - set(kwargs.keys())
            )
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call".format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self.get_hookimpls(), kwargs)

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/hooks.py:286: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.config.PytestPluginManager object at 0x00007f92e228d948>, hook = <_HookCaller 'pytest_runtest_setup'>
methods = [<HookImpl plugin_name='nose', plugin=<module '_pytest.nose' from '/home/matti/miniconda3/envs/pypy37/site-packages/_p...e=None>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x00007f92e1f70b48>>]
kwargs = {'item': <Function test_async_await>}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/manager.py:93: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook = <_HookCaller 'pytest_runtest_setup'>
methods = [<HookImpl plugin_name='nose', plugin=<module '_pytest.nose' from '/home/matti/miniconda3/envs/pypy37/site-packages/_p...e=None>>, <HookImpl plugin_name='logging-plugin', plugin=<_pytest.logging.LoggingPlugin object at 0x00007f92e1f70b48>>]
kwargs = {'item': <Function test_async_await>}

    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
        methods,
        kwargs,
>       firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
    )

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/manager.py:87: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

item = <Function test_async_await>

    def pytest_runtest_setup(item: Item) -> None:
        _update_current_test_var(item, "setup")
>       item.session._setupstate.prepare(item)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/runner.py:150: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.runner.SetupState object at 0x00007f92e1ed84b8>, colitem = <Function test_async_await>

    def prepare(self, colitem) -> None:
        """Setup objects along the collector chain to the test-method."""
    
        # Check if the last collection node has raised an error.
        for col in self.stack:
            if hasattr(col, "_prepare_exc"):
                exc = col._prepare_exc  # type: ignore[attr-defined]
                raise exc
    
        needed_collectors = colitem.listchain()
        for col in needed_collectors[len(self.stack) :]:
            self.stack.append(col)
            try:
>               col.setup()

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/runner.py:449: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Function test_async_await>

    def setup(self) -> None:
        if isinstance(self.parent, Instance):
            self.parent.newinstance()
            self.obj = self._getobj()
>       self._request._fillfixtures()

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/python.py:1647: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <FixtureRequest for <Function test_async_await>>

    def _fillfixtures(self) -> None:
        item = self._pyfuncitem
        fixturenames = getattr(item, "fixturenames", self.fixturenames)
        for argname in fixturenames:
            if argname not in item.funcargs:
>               item.funcargs[argname] = self.getfixturevalue(argname)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/fixtures.py:568: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <FixtureRequest for <Function test_async_await>>, argname = 'xunit_setup_function_fixture_ipykernel.tests.test_async'

    def getfixturevalue(self, argname: str) -> Any:
        """Dynamically run a named fixture function.
    
        Declaring fixtures via function argument is recommended where possible.
        But if you can only decide whether to use another fixture at test
        setup time, you may use this function to retrieve it inside a fixture
        or test function body.
    
        :raises pytest.FixtureLookupError:
            If the given fixture could not be found.
        """
>       fixturedef = self._get_active_fixturedef(argname)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/fixtures.py:581: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <FixtureRequest for <Function test_async_await>>, argname = 'xunit_setup_function_fixture_ipykernel.tests.test_async'

    def _get_active_fixturedef(
        self, argname: str
    ) -> Union["FixtureDef[object]", PseudoFixtureDef[object]]:
        try:
            return self._fixture_defs[argname]
        except KeyError:
            try:
                fixturedef = self._getnextfixturedef(argname)
            except FixtureLookupError:
                if argname == "request":
                    cached_result = (self, [0], None)
                    scope: _Scope = "function"
                    return PseudoFixtureDef(cached_result, scope)
                raise
        # Remove indent to prevent the python3 exception
        # from leaking into the call.
>       self._compute_fixture_value(fixturedef)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/fixtures.py:601: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <FixtureRequest for <Function test_async_await>>
fixturedef = <FixtureDef argname='xunit_setup_function_fixture_ipykernel.tests.test_async' scope='function' baseid='ipykernel/tests/test_async.py'>

    def _compute_fixture_value(self, fixturedef: "FixtureDef[object]") -> None:
        """Create a SubRequest based on "self" and call the execute method
        of the given FixtureDef object.
    
        This will force the FixtureDef object to throw away any previous
        results and compute a new fixture value, which will be stored into
        the FixtureDef object itself.
        """
        # prepare a subrequest object before calling fixture function
        # (latter managed by fixturedef)
        argname = fixturedef.argname
        funcitem = self._pyfuncitem
        scope = fixturedef.scope
        try:
            param = funcitem.callspec.getparam(argname)
        except (AttributeError, ValueError):
            param = NOTSET
            param_index = 0
            has_params = fixturedef.params is not None
            fixtures_not_supported = getattr(funcitem, "nofuncargs", False)
            if has_params and fixtures_not_supported:
                msg = (
                    "{name} does not support fixtures, maybe unittest.TestCase subclass?\n"
                    "Node id: {nodeid}\n"
                    "Function type: {typename}"
                ).format(
                    name=funcitem.name,
                    nodeid=funcitem.nodeid,
                    typename=type(funcitem).__name__,
                )
                fail(msg, pytrace=False)
            if has_params:
                frame = inspect.stack()[3]
                frameinfo = inspect.getframeinfo(frame[0])
                source_path = py.path.local(frameinfo.filename)
                source_lineno = frameinfo.lineno
                rel_source_path = source_path.relto(funcitem.config.rootdir)
                if rel_source_path:
                    source_path_str = rel_source_path
                else:
                    source_path_str = str(source_path)
                msg = (
                    "The requested fixture has no parameter defined for test:\n"
                    "    {}\n\n"
                    "Requested fixture '{}' defined in:\n{}"
                    "\n\nRequested here:\n{}:{}".format(
                        funcitem.nodeid,
                        fixturedef.argname,
                        getlocation(fixturedef.func, funcitem.config.rootdir),
                        source_path_str,
                        source_lineno,
                    )
                )
                fail(msg, pytrace=False)
        else:
            param_index = funcitem.callspec.indices[argname]
            # If a parametrize invocation set a scope it will override
            # the static scope defined with the fixture function.
            paramscopenum = funcitem.callspec._arg2scopenum.get(argname)
            if paramscopenum is not None:
                scope = scopes[paramscopenum]
    
        subrequest = SubRequest(
            self, scope, param, param_index, fixturedef, _ispytest=True
        )
    
        # Check if a higher-level scoped fixture accesses a lower level one.
        subrequest._check_scope(argname, self.scope, scope)
        try:
            # Call the fixture function.
>           fixturedef.execute(request=subrequest)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/fixtures.py:687: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <FixtureDef argname='xunit_setup_function_fixture_ipykernel.tests.test_async' scope='function' baseid='ipykernel/tests/test_async.py'>
request = <SubRequest 'xunit_setup_function_fixture_ipykernel.tests.test_async' for <Function test_async_await>>

    def execute(self, request: SubRequest) -> _FixtureValue:
        # Get required arguments and register our own finish()
        # with their finalization.
        for argname in self.argnames:
            fixturedef = request._get_active_fixturedef(argname)
            if argname != "request":
                # PseudoFixtureDef is only for "request".
                assert isinstance(fixturedef, FixtureDef)
                fixturedef.addfinalizer(functools.partial(self.finish, request=request))
    
        my_cache_key = self.cache_key(request)
        if self.cached_result is not None:
            # note: comparison with `==` can fail (or be expensive) for e.g.
            # numpy arrays (#6497).
            cache_key = self.cached_result[1]
            if my_cache_key is cache_key:
                if self.cached_result[2] is not None:
                    _, val, tb = self.cached_result[2]
                    raise val.with_traceback(tb)
                else:
                    result = self.cached_result[0]
                    return result
            # We have a previous but differently parametrized fixture instance
            # so we need to tear it down before creating a new one.
            self.finish(request)
            assert self.cached_result is None
    
        hook = self._fixturemanager.session.gethookproxy(request.node.fspath)
>       result = hook.pytest_fixture_setup(fixturedef=self, request=request)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/fixtures.py:1072: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_HookCaller 'pytest_fixture_setup'>, args = ()
kwargs = {'fixturedef': <FixtureDef argname='xunit_setup_function_fixture_ipykernel.tests.test_async' scope='function' baseid='...y'>, 'request': <SubRequest 'xunit_setup_function_fixture_ipykernel.tests.test_async' for <Function test_async_await>>}
notincall = set()

    def __call__(self, *args, **kwargs):
        if args:
            raise TypeError("hook calling supports only keyword arguments")
        assert not self.is_historic()
        if self.spec and self.spec.argnames:
            notincall = (
                set(self.spec.argnames) - set(["__multicall__"]) - set(kwargs.keys())
            )
            if notincall:
                warnings.warn(
                    "Argument(s) {} which are declared in the hookspec "
                    "can not be found in this hook call".format(tuple(notincall)),
                    stacklevel=2,
                )
>       return self._hookexec(self, self.get_hookimpls(), kwargs)

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/hooks.py:286: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <_pytest.config.PytestPluginManager object at 0x00007f92e228d948>, hook = <_HookCaller 'pytest_fixture_setup'>
methods = [<HookImpl plugin_name='fixtures', plugin=<module '_pytest.fixtures' from '/home/matti/miniconda3/envs/pypy37/site-pac...ly', plugin=<module '_pytest.setuponly' from '/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/setuponly.py'>>]
kwargs = {'fixturedef': <FixtureDef argname='xunit_setup_function_fixture_ipykernel.tests.test_async' scope='function' baseid='...y'>, 'request': <SubRequest 'xunit_setup_function_fixture_ipykernel.tests.test_async' for <Function test_async_await>>}

    def _hookexec(self, hook, methods, kwargs):
        # called from all hookcaller instances.
        # enable_tracing will set its own wrapping function at self._inner_hookexec
>       return self._inner_hookexec(hook, methods, kwargs)

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/manager.py:93: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

hook = <_HookCaller 'pytest_fixture_setup'>
methods = [<HookImpl plugin_name='fixtures', plugin=<module '_pytest.fixtures' from '/home/matti/miniconda3/envs/pypy37/site-pac...ly', plugin=<module '_pytest.setuponly' from '/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/setuponly.py'>>]
kwargs = {'fixturedef': <FixtureDef argname='xunit_setup_function_fixture_ipykernel.tests.test_async' scope='function' baseid='...y'>, 'request': <SubRequest 'xunit_setup_function_fixture_ipykernel.tests.test_async' for <Function test_async_await>>}

    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
        methods,
        kwargs,
>       firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
    )

/home/matti/miniconda3/envs/pypy37/site-packages/pluggy/manager.py:87: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

fixturedef = <FixtureDef argname='xunit_setup_function_fixture_ipykernel.tests.test_async' scope='function' baseid='ipykernel/tests/test_async.py'>
request = <SubRequest 'xunit_setup_function_fixture_ipykernel.tests.test_async' for <Function test_async_await>>

    def pytest_fixture_setup(
        fixturedef: FixtureDef[_FixtureValue], request: SubRequest
    ) -> _FixtureValue:
        """Execution of fixture setup."""
        kwargs = {}
        for argname in fixturedef.argnames:
            fixdef = request._get_active_fixturedef(argname)
            assert fixdef.cached_result is not None
            result, arg_cache_key, exc = fixdef.cached_result
            request._check_scope(argname, request.scope, fixdef.scope)
            kwargs[argname] = result
    
        fixturefunc = resolve_fixture_function(fixturedef, request)
        my_cache_key = fixturedef.cache_key(request)
        try:
>           result = call_fixture_func(fixturefunc, request, kwargs)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/fixtures.py:1126: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

fixturefunc = <function Module._inject_setup_function_fixture.<locals>.xunit_setup_function_fixture at 0x000056409deba660>
request = <SubRequest 'xunit_setup_function_fixture_ipykernel.tests.test_async' for <Function test_async_await>>
kwargs = {'request': <SubRequest 'xunit_setup_function_fixture_ipykernel.tests.test_async' for <Function test_async_await>>}

    def call_fixture_func(
        fixturefunc: "_FixtureFunc[_FixtureValue]", request: FixtureRequest, kwargs
    ) -> _FixtureValue:
        if is_generator(fixturefunc):
            fixturefunc = cast(
                Callable[..., Generator[_FixtureValue, None, None]], fixturefunc
            )
            generator = fixturefunc(**kwargs)
            try:
>               fixture_result = next(generator)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/fixtures.py:925: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

request = <SubRequest 'xunit_setup_function_fixture_ipykernel.tests.test_async' for <Function test_async_await>>

    @fixtures.fixture(
        autouse=True,
        scope="function",
        # Use a unique name to speed up lookup.
        name=f"xunit_setup_function_fixture_{self.obj.__name__}",
    )
    def xunit_setup_function_fixture(request) -> Generator[None, None, None]:
        if request.instance is not None:
            # in this case we are bound to an instance, so we need to let
            # setup_method handle this
            yield
            return
        if setup_function is not None:
>           _call_with_optional_argument(setup_function, request.function)

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/python.py:567: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

func = <function setup_function at 0x000056409deba980>, arg = <function test_async_await at 0x000056409deba700>

    def _call_with_optional_argument(func, arg) -> None:
        """Call the given function with the given argument if func accepts one argument, otherwise
        calls func without arguments."""
        arg_count = func.__code__.co_argcount
        if inspect.ismethod(func):
            arg_count -= 1
        if arg_count:
            func(arg)
        else:
>           func()

/home/matti/miniconda3/envs/pypy37/site-packages/_pytest/python.py:741: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def setup_function():
        """start the global kernel (if it isn't running) and return its client"""
        global KM, KC
>       KM, KC = start_new_kernel()

ipykernel/tests/test_async.py:20: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

kwargs = {'stderr': -2}

    def start_new_kernel(**kwargs):
        """start a new kernel, and return its Manager and Client
    
        Integrates with our output capturing for tests.
        """
        kwargs['stderr'] = STDOUT
        try:
            kwargs['stdout'] = nose.iptest_stdstreams_fileno()
        except AttributeError:
            pass
>       return manager.start_new_kernel(startup_timeout=STARTUP_TIMEOUT, **kwargs)

ipykernel/tests/utils.py:36: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

startup_timeout = 60, kernel_name = 'python', kwargs = {'stderr': -2}
km = <jupyter_client.manager.KernelManager object at 0x000056409e58c560>
kc = <jupyter_client.blocking.client.BlockingKernelClient object at 0x000056409e615520>

    def start_new_kernel(startup_timeout=60, kernel_name='python', **kwargs):
        """Start a new kernel, and return its Manager and Client"""
        km = KernelManager(kernel_name=kernel_name)
        km.start_kernel(**kwargs)
        kc = km.client()
        kc.start_channels()
        try:
>           kc.wait_for_ready(timeout=startup_timeout)

/home/matti/miniconda3/envs/pypy37/site-packages/jupyter_client/manager.py:787: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <jupyter_client.blocking.client.BlockingKernelClient object at 0x000056409e615520>, timeout = 60

    def wait_for_ready(self, timeout=None):
        """Waits for a response when a client is blocked
    
        - Sets future time for timeout
        - Blocks on shell channel until a message is received
        - Exit if the kernel has died
        - If client times out before receiving a message from the kernel, send RuntimeError
        - Flush the IOPub channel
        """
        if timeout is None:
            abs_timeout = float('inf')
        else:
            abs_timeout = time.time() + timeout
    
        from ..manager import KernelManager
        if not isinstance(self.parent, KernelManager):
            # This Client was not created by a KernelManager,
            # so wait for kernel to become responsive to heartbeats
            # before checking for kernel_info reply
            while not self.is_alive():
                if time.time() > abs_timeout:
                    raise RuntimeError("Kernel didn't respond to heartbeats in %d seconds and timed out" % timeout)
                time.sleep(0.2)
    
        # Wait for kernel info reply on shell channel
        while True:
            self.kernel_info()
            try:
>               msg = self.shell_channel.get_msg(block=True, timeout=1)

/home/matti/miniconda3/envs/pypy37/site-packages/jupyter_client/blocking/client.py:96: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <jupyter_client.blocking.channels.ZMQSocketChannel object at 0x000056409e63d2b8>, block = True, timeout = 1000

    def get_msg(self, block=True, timeout=None):
        """ Gets a message if there is one that is ready. """
        if block:
            if timeout is not None:
                timeout *= 1000  # seconds to ms
>           ready = self.socket.poll(timeout)

/home/matti/miniconda3/envs/pypy37/site-packages/jupyter_client/blocking/channels.py:47: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <zmq.sugar.socket.Socket object at 0x000056409e63d018>, timeout = 1000, flags = 1

    def poll(self, timeout=None, flags=POLLIN):
        """Poll the socket for events.
        See :class:`Poller` to wait for multiple sockets at once.
    
        Parameters
        ----------
        timeout : int [default: None]
            The timeout (in milliseconds) to wait for an event. If unspecified
            (or specified None), will wait forever for an event.
        flags : int [default: POLLIN]
            POLLIN, POLLOUT, or POLLIN|POLLOUT. The event flags to poll for.
    
        Returns
        -------
        event_mask : int
            The poll event mask (POLLIN, POLLOUT),
            0 if the timeout was reached without an event.
        """
    
        if self.closed:
            raise ZMQError(ENOTSUP)
    
        p = self._poller_class()
        p.register(self, flags)
>       evts = dict(p.poll(timeout))

/home/matti/miniconda3/envs/pypy37/site-packages/zmq/sugar/socket.py:810: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <zmq.sugar.poll.Poller object at 0x000056409e656b48>, timeout = 1000

    def poll(self, timeout: Optional[int] = None):
        """Poll the registered 0MQ or native fds for I/O.
    
        If there are currently events ready to be processed, this function will return immediately.
        Otherwise, this function will return as soon the first event is available or after timeout
        milliseconds have elapsed.
    
        Parameters
        ----------
        timeout : float, int
            The timeout in milliseconds. If None, no `timeout` (infinite). This
            is in milliseconds to be compatible with ``select.poll()``.
    
        Returns
        -------
        events : list of tuples
            The list of events that are ready to be processed.
            This is a list of tuples of the form ``(socket, event_mask)``, where the 0MQ Socket
            or integer fd is the first element, and the poll event mask (POLLIN, POLLOUT) is the second.
            It is common to call ``events = dict(poller.poll())``,
            which turns the list of tuples into a mapping of ``socket : event_mask``.
        """
        if timeout is None or timeout < 0:
            timeout = -1
        elif isinstance(timeout, float):
            timeout = int(timeout)
>       return zmq_poll(self.sockets, timeout=timeout)

/home/matti/miniconda3/envs/pypy37/site-packages/zmq/sugar/poll.py:105: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

sockets = [(<zmq.sugar.socket.Socket object at 0x000056409e63d018>, 1)], timeout = 1000

    def zmq_poll(sockets, timeout):
        cffi_pollitem_list = []
        low_level_to_socket_obj = {}
        from zmq import Socket
    
        for item in sockets:
            if isinstance(item[0], Socket):
                low_level_to_socket_obj[item[0]._zmq_socket] = item
                cffi_pollitem_list.append(_make_zmq_pollitem(item[0], item[1]))
            else:
                if not isinstance(item[0], int):
                    # not an FD, get it from fileno()
                    item = (item[0].fileno(), item[1])
                low_level_to_socket_obj[item[0]] = item
                cffi_pollitem_list.append(_make_zmq_pollitem_fromfd(item[0], item[1]))
        items = ffi.new('zmq_pollitem_t[]', cffi_pollitem_list)
        list_length = ffi.cast('int', len(cffi_pollitem_list))
        while True:
            c_timeout = ffi.cast('long', timeout)
            start = monotonic()
>           rc = C.zmq_poll(items, list_length, c_timeout)
E           KeyboardInterrupt

/home/matti/miniconda3/envs/pypy37/site-packages/zmq/backend/cffi/_poll.py:56: KeyboardInterrupt
====================================================== 5 warnings in 23.57s ======================================================

```

--- a/.github/workflows/wheels.yml
+++ b/.github/workflows/wheels.yml
@@ -101,7 +101,7 @@ jobs:
           - os: ubuntu-20.04
             name: manylinux2010
             cibw:
-              build: "cp38* cp39*"
+              build: "cp38* cp39* pp3*"
               manylinux_image: manylinux2010
 
           - os: ubuntu-20.04

--- a/docs/source/changelog.rst
+++ b/docs/source/changelog.rst
@@ -9,6 +9,12 @@ Changes in PyZMQ
 This is a coarse summary of changes in pyzmq versions.
 For a full changelog, consult the `git log <https://github.com/zeromq/pyzmq/commits>`_.
 
+22.0.1
+======
+
+- Fix type of ``Frame.bytes`` for non-copying recvs with CFFI backend (regression in 21.0)
+- Add manylinux wheels for pypy
+
 22.0.0
 ======
 

--- a/zmq/backend/cffi/message.py
+++ b/zmq/backend/cffi/message.py
@@ -21,7 +21,14 @@ try:
 except ImportError:
     maybe_bufferable = object
 
-_content = lambda x: x.tobytes() if type(x) == memoryview else x
+
+def _content(obj):
+    """Return content of obj as bytes"""
+    if type(obj) is bytes:
+        return obj
+    if not isinstance(obj, memoryview):
+        obj = memoryview(obj)
+    return obj.tobytes()
 
 
 def _check_rc(rc):
@@ -44,6 +51,7 @@ class Frame(maybe_bufferable):
     closed = False
     more = False
     _buffer = None
+    _bytes = None
     _failed_init = False
     tracker_event = None
     zmq_msg = None
@@ -70,6 +78,9 @@ class Frame(maybe_bufferable):
             return
 
         self._data = data
+        if type(data) is bytes:
+            # avoid unnecessary copy on .bytes access
+            self._bytes = data
 
         self._buffer = memoryview(data)
         c_data = ffi.from_buffer(self._buffer)
@@ -160,9 +171,9 @@ class Frame(maybe_bufferable):
 
     @property
     def bytes(self):
-        if self._data is None:
-            self._buffer_from_zmq_msg()
-        return _content(self._data)
+        if self._bytes is None:
+            self._bytes = self.buffer.tobytes()
+        return self._bytes
 
     def __len__(self):
         return self.buffer.nbytes

--- a/zmq/tests/test_message.py
+++ b/zmq/tests/test_message.py
@@ -301,6 +301,7 @@ class TestFrame(BaseZMQTestCase):
                 self.assertEqual(b, null)
                 self.assertEqual(mb, null)
                 self.assertEqual(m2.bytes, ff)
+                assert type(m2.bytes) is bytes
 
     def test_buffer_numpy(self):
         """test non-copying numpy array messages"""

